src/Common/src/CoreWCF/Runtime/AsyncLock.cs

```diff
@@ -51,6 +51,21 @@ public async Task<IDisposable> TakeLockAsync()
             return _semaphoreRelease;
         }
 
+        public Task<IDisposable> TakeLockAsync(TimeSpan timeout)
+        {
+            if (timeout == Timeout.InfiniteTimeSpan || timeout < TimeSpan.Zero || timeout == TimeSpan.MaxValue)
```

Comment:
Should we impose same check for TakeLock(int timeout), when timeout is 0 ? Or in both cases throw exception if not within certain range..

Reply:
A value of 0 means immediately timed out which means it will throw an OperationCancelledException from the method the token is passed to. We could short-circuit that here but it's going to be a rare case and adding the redundant check when it's checked later will just add unnecessary overhead.